lib/puppet/type/exec.rb

```diff
@@ -204,7 +204,7 @@ def value=(*values)
       validate do |user|
         if Puppet.features.microsoft_windows?
           self.fail "Unable to execute commands as other users on Windows"
-        elsif !Puppet.features.root? && resource.current_username() != user
+        elsif Facter.value(:id) != 'root' && Facter.value(:id) != user
```

Comment:
@mcanevet I'm not understanding why this is necessary. The validate block is executed on the agent when it converts the resource catalog (transmitted via REST) to the RAL catalog. Here is a trace leading up to this:

<pre>
/Users/josh/work/puppet/lib/puppet/provider/exec.rb:89:in `validatecmd'
/Users/josh/work/puppet/lib/puppet/type/exec.rb:500:in `block (2 levels) in <module:Puppet>'
/Users/josh/work/puppet/lib/puppet/type.rb:2311:in `initialize'
/Users/josh/work/puppet/lib/puppet/resource.rb:377:in `new'
/Users/josh/work/puppet/lib/puppet/resource.rb:377:in `to_ral'
/Users/josh/work/puppet/lib/puppet/resource/catalog.rb:501:in `block in to_catalog'
/Users/josh/work/puppet/lib/puppet/resource/catalog.rb:493:in `each'
/Users/josh/work/puppet/lib/puppet/resource/catalog.rb:493:in `to_catalog'
/Users/josh/work/puppet/lib/puppet/resource/catalog.rb:396:in `to_ral'
/Users/josh/work/puppet/lib/puppet/configurer.rb:76:in `convert_catalog'
</pre>

I suspect that you are seeing an issue because your test is both the agent and the master, and is not stubbing `Puppet.features.stubs(:root?).returns(true)`.